tests/rados/test_pool_osd_recovery.py

```diff
@@ -0,0 +1,337 @@
+    Returns:
+        1 -> Fail, 0 -> Pass
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+    pool_configs = config["pool_configs"]
+    pool_configs_path = config["pool_configs_path"]
+    osd_nodes = ceph_cluster.get_nodes(role="osd")
+
+    with open(pool_configs_path, "r") as fd:
+        pool_conf_file = yaml.safe_load(fd)
+
+    pools = []
+    for i in pool_configs:
+        pool = pool_conf_file[i["type"]][i["conf"]]
+        create_given_pool(rados_obj, pool)
+        pools.append(pool["pool_name"])
+
+    log.info(f"Created {len(pools)} pools for testing. pools : {pools}")
+
+    # Checking cluster health before starting the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info(
+        "Completed health cehck of the cluster after test pool creation. Cluster health good!!"
+    )
+
+    log.info("\n\n---- Starting All workflows ----\n\n")
+    # rebooting one OSD and checking cluster health
+    log.info(
+        "---- Starting workflow ----\n---- 1. Reboot of single osd and health check ----"
+    )
+    for pool in pools:
+        target_osd = rados_obj.get_pg_acting_set(pool_name=pool)[0]
+        log.debug(f"Rebooting OSD : {target_osd} and checking health status")
+        if not rados_obj.change_osd_state(action="restart", target=target_osd):
+            log.error(f"Unable to restart the OSD : {target_osd}")
+            raise Exception("Execution error")
+
+        # Waiting for recovery to post OSD reboot
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.debug(
+            "PG's are active + clean post OSD reboot, proceeding to restart next OSD"
+        )
+
+    log.info("All the planned primary OSD reboots have completed")
+    # Checking cluster health after the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info("----Completed workflows 1. Reboot of single osd and health check----")
+
+    log.info("---- Starting workflow ----\n---- 2. Rolling Reboot OSD hosts ----")
+    for nodes in osd_nodes:
+        log.info(f"Rebooting node : {nodes.hostname}")
+        nodes.exec_command(sudo=True, cmd="reboot", long_running=True)
+
+        # Sleeping for 10 seconds and starting verification.
+        time.sleep(10)
+
+        # Waiting for recovery to post OSD host reboot
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.info(f"PG's are active + clean post reboot of host {nodes.hostname}")
+
+    log.info("Completed reboot of all the OSD hosts, Checking cluster health status")
+    # Checking cluster health after the test
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info(
+        "---- Completed workflows 2. Rolling Reboot OSD hosts and health check ----"
+    )
+
+    log.info(
+        "---- Starting workflow ----\n---- 3. Stopping and starting OSD daemons ----"
```

Comment:
As per the test case mentioned, both scenario performs the same scenarios. Reboot of all the Primary PG OSD followed by the remaining acting set OSD. Step 3 & 4 in the TC CEPH-11032. Please modify the TC if required.

Reply:
Reboot is different from start and stop. 

In reboot OSDs are generally not marked out. In stop scenario, we are waiting for the data in the down OSD to get copied into other OSD.

In reboot, data is not usually copied, but the delta of the data written during downtime is copied.

Of course this is not verified explicitly, but rather we wait for PGs to get clean with a OSD down.

Let me know what is the same scenario here, or if something needs modification